File: user_app/test/clinical_tool_migration_wave1_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:user_app/features/calculators/data/models/clinical_tool_definition.dart';
import 'package:user_app/features/calculators/domain/clinical_tool_evaluator.dart';

void main() {
  const files = [
    'apgar-score-calculator',
    'blood-pressure-assessment',
    'bmi-calculator',
    'glasgow-coma-scale',
    'fluid-balance-calculator',
    'pain-assessment-scale',
    'pregnancy-due-date-calculator',
    'dehydration-assessment',
    'pediatric-fever-management',
    'wound-assessment-tool',
    'cardiac-risk-assessment',
    'emergency-triage-assessment',
    'immunization-schedule-checker',
    'medication-dosage-calculator',
  ];

  for (final file in files) {
    final envelope = Map<String, dynamic>.from(
      jsonDecode(
            File(
              '../clinical-tools/migrations/v1/definitions/$file.json',
            ).readAsStringSync(),
          )
          as Map,
    );
    final definitionJson = Map<String, dynamic>.from(
      envelope['definition'] as Map,
    );
    final definition = ClinicalToolDefinition.fromJson(definitionJson);
    final parity = Map<String, dynamic>.from(
      jsonDecode(
            File(
              '../clinical-tools/migrations/v1/parity/$file.json',
            ).readAsStringSync(),
          )
          as Map,
    );

    test('$file remains explicitly unapproved', () {
      expect(parity['status'], 'changes_required');
      expect(parity['reviewer_id'], isNull);
      expect(parity['reviewed_at'], isNull);
    });

    for (final caseValue in definitionJson['test_cases'] as List<dynamic>) {
      final fixture = Map<String, dynamic>.from(caseValue as Map);
      test('$file: ${fixture['key']}', () {
        final fixedNow = fixture['fixed_now'] == null
            ? null
            : DateTime.parse(fixture['fixed_now'].toString());
        final result = ClinicalToolEvaluator(fixedNow: fixedNow).evaluate(
          definition,
          Map<String, Object?>.from(fixture['inputs'] as Map),
        );
        final expected = Map<String, dynamic>.from(fixture['expected'] as Map);
        final outputKeys = definition.outputs
            .map((output) => output.key)
            .toSet();
        final tolerance =
            (fixture['numeric_tolerance'] as num?)?.toDouble() ?? 0;
        for (final entry in expected.entries) {
          if (!outputKeys.contains(entry.key)) continue;
          if (entry.value is num) {
            expect(
              result.values[entry.key],
              isA<num>().having(
                (value) => value.toDouble(),
                entry.key,
                closeTo((entry.value as num).toDouble(), tolerance),
              ),
            );
          } else {
            expect(result.values[entry.key], entry.value);
          }
        }
        final interpretationKeys = expected['interpretations'] as List?;
        if (interpretationKeys != null) {
          expect(result.interpretation?.key, interpretationKeys.first);
        }
        final warningKeys = expected['warnings'] as List?;
        if (warningKeys != null) {
          expect(result.warnings.map((warning) => warning.key), warningKeys);
        }
      });
    }
  }
}

File: dashboard/lib/clinical-tool-migration-wave1.test.ts
import { readFileSync } from "node:fs"
import { join } from "node:path"
import { describe, expect, it } from "vitest"

import type { ClinicalToolDefinition } from "@/services/clinical-tool.service"
import { previewClinicalTool } from "./clinical-tool-evaluator"

type MigrationEnvelope = {
  definition: ClinicalToolDefinition & {
    test_cases: Array<{ key: string; inputs: Record<string, unknown>; expected: Record<string, unknown>; fixed_now?: string; numeric_tolerance?: number }>
  }
}

const files = [
  "apgar-score-calculator", "blood-pressure-assessment", "bmi-calculator", "glasgow-coma-scale",
  "fluid-balance-calculator", "pain-assessment-scale", "pregnancy-due-date-calculator",
  "dehydration-assessment", "pediatric-fever-management", "wound-assessment-tool",
  "cardiac-risk-assessment", "emergency-triage-assessment", "immunization-schedule-checker",
  "medication-dosage-calculator",
]

describe("clinical-tool migration review drafts", () => {
  for (const file of files) {
    const envelope = JSON.parse(readFileSync(join(process.cwd(), "..", "clinical-tools", "migrations", "v1", "definitions", `${file}.json`), "utf8")) as MigrationEnvelope
    const parity = JSON.parse(readFileSync(join(process.cwd(), "..", "clinical-tools", "migrations", "v1", "parity", `${file}.json`), "utf8")) as { status: string; reviewer_id: string | null; reviewed_at: string | null }

    it(`${file} remains explicitly unapproved`, () => {
      expect(parity.status).toBe("changes_required")
      expect(parity.reviewer_id).toBeNull()
      expect(parity.reviewed_at).toBeNull()
    })

    for (const fixture of envelope.definition.test_cases) {
      it(`${file}: ${fixture.key}`, () => {
        const result = previewClinicalTool(envelope.definition, fixture.inputs, { fixedNow: fixture.fixed_now })
        const outputKeys = new Set(envelope.definition.outputs.map((output) => output.key))
        const tolerance = fixture.numeric_tolerance ?? 0
        for (const [key, expectedValue] of Object.entries(fixture.expected)) {
          if (outputKeys.has(key)) {
            if (typeof expectedValue === "number") {
              if (tolerance > 0) expect(Math.abs(Number(result.values[key]) - expectedValue)).toBeLessThanOrEqual(tolerance)
              else expect(Number(result.values[key])).toBeCloseTo(expectedValue, 10)
            }
            else expect(result.values[key]).toEqual(expectedValue)
          }
        }
        const expectedInterpretations = fixture.expected.interpretations as string[] | undefined
        if (expectedInterpretations) {
          const expected = envelope.definition.interpretations.find((item) => item.key === expectedInterpretations[0])
          expect(result.interpretation).toBe(expected?.label)
          expect(result.recommendations).toEqual(expected?.recommendations ?? [])
        }
        const expectedWarningKeys = fixture.expected.warnings as string[] | undefined
        if (expectedWarningKeys) {
          const expectedWarnings = expectedWarningKeys.map((key) => envelope.definition.warnings?.find((item) => item.key === key)?.text)
          expect(result.warnings).toEqual(expectedWarnings)
        }
      })
    }
  }
})
